configuration-versioning/README.md

```diff
@@ -0,0 +1,137 @@
+    packages_to_install = fields.String(missing=None)
+    owner = fields.String(missing=None)
+    project = fields.String(missing=None)
+```
+
+and the config object would have all the fields and the code could handle both
+
+- from the coding perspective this would not be a problem - the version would be detected either from whole config
+  or directly from the `image_builder_metadata`
+- but this may be confusing to users - versioning some configuration field apart from the whole config,
+  on the other hand it would be independent on the rest of the package config
+
+## Implementation details
+
+### Versioning
+
+- for each backwards incompatible change, we would bump the version
+- version would default to 1 (if not present)
+- this would be properly documented, we can get inspired by Docker,
+  [see](https://docs.docker.com/compose/compose-file/compose-versioning/#versioning), the documentation
+  would also include Packit version that starts supporting the particular configuration schema
+
+### Having more schemas vs having one schema
+
+#### Having more schemas
+
+- maintaining more versions of schema for whole package config should not be needed
+  - the schemas would be mostly have the same content, only with difference in few fields
+- this approach could be used for smaller parts of the configuration, e.g.
+  creating nested schema (as described for Example 1)
+- the coding around would be pretty straightforward, detect the version
+  and use the correct schema for loading/dumping accordingly
+
+#### Having one schema
+
+- in case of Example 2, the schema itself doesn't change
+- we could use the `pre_load` and `pre_dump` (we already use these) to migrate the data:
+
+```python
+    @staticmethod
+    def propagate_options_to_jobs(data, **_):
+        """
+        add all the fields (except for jobs) to every job so we can process only jobconfig in p-s
+        """
+        if not data:  # data is None when .packit.yaml is empty
+            return data
+        for job in data.get("jobs", []):
+            for k, v in data.items():
+                if k == "jobs":
+                    # overriding jobs doesn't make any sense
+                    continue
+                if self.version == 1:
+                    job.setdefault(k, v)
+                elif self.version == 2:
+                     # merge the values from top-level and the particular job
+                ...
+
+        return data
+```
+
+- we would need to make sure that load-dump-load results in same representation
+
+## Plan
+
+- decide whether/to which scope we want to do this in relation to [this issue](https://github.com/packit/research/issues/159)
+  - if we would introduce some change that would break the compatibility, we would open PRs for each repository that
+    uses the particular configuration field
+  - this could not be sufficient for all use cases
+- decide whether we want to version whole config or just have versioned fields or use combination
```

Comment:
Versioning jobs or sections sound appealing although I am concerned about usability for users. Even right now, many of our users still use deprecated options (including us!) so we should minimize the number of changes needed to be done to packit.yaml.

Once Hunor finishes his config changes related to monorepo, I wonder how will those changes impact this research.

Reply:
> Versioning jobs or sections sound appealing although I am concerned about usability for users.

I agree, it could be simpler from an implementation pov, but very confusing to users.

Reply:
Versioning sections of the schema has the strong potential of creating a very confusing and difficult to follow environment both for developers and users. I'm strongly against going in this direction.